src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/PreferIsKindAnalyzer.cs

```diff
@@ -0,0 +1,83 @@
+            DiagnosticCategory.MicrosoftCodeAnalysisPerformance,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            description: s_localizableDescription,
+            helpLinkUri: null,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                // Kind() methods
+                //  Microsoft.CodeAnalysis.CSharp.CSharpExtensions.Kind
+                //  Microsoft.CodeAnalysis.VisualBasic.VisualBasicExtensions.Kind
+                //
+                // IsKind() methods
+                //  Microsoft.CodeAnalysis.CSharpExtensions.IsKind
+                //  Microsoft.CodeAnalysis.VisualBasicExtensions.IsKind
+
+                Dictionary<INamedTypeSymbol, INamedTypeSymbol> containingTypeMap = new Dictionary<INamedTypeSymbol, INamedTypeSymbol>();
+                if (context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisCSharpCSharpExtensions) is { } csharpKindExtensions
+                    && context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisCSharpExtensions) is { } csharpIsKindExtensions)
+                {
+                    containingTypeMap[csharpKindExtensions] = csharpIsKindExtensions;
+                }
+
+                if (context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisVisualBasicVisualBasicExtensions) is { } vbKindExtensions
+                    && context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisVisualBasicExtensions) is { } vbIsKindExtensions)
+                {
+                    containingTypeMap[vbKindExtensions] = vbIsKindExtensions;
+                }
+
+                if (containingTypeMap.Count > 0)
+                {
+                    context.RegisterOperationAction(context => HandleBinaryOperation(context, containingTypeMap), OperationKind.Binary);
+                }
+            });
+        }
+
+        private static void HandleBinaryOperation(OperationAnalysisContext context, Dictionary<INamedTypeSymbol, INamedTypeSymbol> containingTypeMap)
+        {
+            var operation = (IBinaryOperation)context.Operation;
+            if (operation.OperatorKind != BinaryOperatorKind.Equals
+                && operation.OperatorKind != BinaryOperatorKind.NotEquals)
+            {
+                return;
+            }
+
+            if (operation.LeftOperand.WalkDownConversion() is IInvocationOperation { TargetMethod: { Name: "Kind", ContainingType: var containingType } }
```

Comment:
maybe assert something aobut hte argument count?

Reply:
If we verify the parameter count, the pattern becomes:

```
IInvocationOperation { TargetMethod: { Name: "Kind", ContainingType: var containingType } and ({ ReducedFrom: null, Parameters: { Length: 1 } } or { ReducedFrom: not null, Parameters: { IsEmpty: true } }) }
```

I wrote the simpler form on the assumption that the symbols with this name inside these well known types only have the expected signatures.

Reply:
Wfm